File: api/src/vibecanvas_api/schemas/llm_credentials.py
"""LLM credential schemas for the API Management Center.

THE SECRECY CONTRACT. ``api_key`` (and the model_name / api_url config) are
private to the owner; they are surfaced through layered out-shapes so a "list"
or "picker" surface (PromptNode / agent — a LATER phase) NEVER sees secret
material:

  - ``CredentialPublicOut``  — id, name, description, provider, timestamps.
    The ONLY fields any list/public surface returns. NO model_name / api_url /
    api_key. (``provider`` IS public — it lets a picker show a provider badge.)
  - ``CredentialOwnerOut``   — the owner's management/edit view: adds
    model_name + api_url + an ``api_key_set: bool`` flag. STILL no plaintext key.

Keys are write-only at the API boundary and stored through the envelope-
encrypted SecretService; ordinary management APIs never reveal plaintext.
"""
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

from vibecanvas_api.schemas.access import ResourceAccessOut

# ----------------------------------------------------------------- out shapes


class CredentialPublicOut(BaseModel):
    """Public / list view. NEVER includes model_name, api_url, or api_key."""

    id: str
    name: str
    description: Optional[str] = None
    provider: str
    connection_kind: Literal["manual", "openrouter_oauth"] = "manual"
    runtime_scope: str
    model_context_tokens: Optional[int] = None
    created_at: datetime
    updated_at: datetime
    access: ResourceAccessOut | None = None


class CredentialOwnerOut(BaseModel):
    """Owner management / edit view. Exposes the non-secret config
    (model_name, api_url) plus an ``api_key_set`` flag — but NEVER the
    plaintext key."""

    id: str
    name: str
    description: Optional[str] = None
    provider: str
    connection_kind: Literal["manual", "openrouter_oauth"] = "manual"
    runtime_scope: str
    model_name: str
    model_context_tokens: Optional[int] = None
    api_url: Optional[str] = None
    # Optional outbound proxy. Owner-only (like api_url) — may carry
    # ``user:pass@host``, so it is NEVER on CredentialPublicOut.
    proxy: Optional[str] = None
    api_key_set: bool
    created_at: datetime
    updated_at: datetime
    access: ResourceAccessOut | None = None


class CredentialConnectionTestOut(BaseModel):
    """Sanitized connection-test result. Provider response bodies and secret
    material never cross the API boundary."""

    ok: bool
    outcome: Literal[
        "connected",
        "credentials_rejected",
        "endpoint_rejected",
        "unreachable",
    ]
    latency_ms: int
    upstream_status: Optional[int] = None


class OpenRouterModelOut(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    name: str
    description: str = ""
    context_length: int | None = None
    input_modalities: list[str] = Field(default_factory=list)
    output_modalities: list[str] = Field(default_factory=list)
    supports_tools: bool = False
    supports_web_search: bool = False
    supported_reasoning_efforts: list[str] = Field(default_factory=list)
    default_reasoning_effort: str | None = None
    pricing: dict[str, str | None] = Field(default_factory=dict)
    available: bool = True


class OpenRouterConnectionOut(BaseModel):
    connected: bool
    credential_id: str | None = None
    models: list[OpenRouterModelOut] = Field(default_factory=list)
    catalog_refreshed_at: datetime | None = None
    catalog_stale: bool = False
    error_code: str | None = None


class OpenRouterStartOut(BaseModel):
    authorization_url: str
    expires_at: datetime


class OpenRouterCallbackIn(BaseModel):
    model_config = ConfigDict(extra="forbid")

    code: str = Field(min_length=1, max_length=4000)
    state: str = Field(min_length=32, max_length=512)


# ----------------------------------------------------------------- in shapes


class CredentialCreate(BaseModel):
    """Create body. ``extra='forbid'`` blocks a client smuggling
    tenant_id / user_id / id / enabled (G4b trust boundary)."""

    model_config = ConfigDict(extra="forbid")

    name: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = Field(default=None, max_length=2000)
    provider: str = Field(..., min_length=1, max_length=100)
    runtime_scope: str = Field(default="codex", pattern=r"^[a-z][a-z0-9_-]{0,63}$")
    model_name: str = Field(..., min_length=1, max_length=200)
    model_context_tokens: Optional[int] = Field(default=None, gt=0)
    api_url: Optional[str] = Field(default=None, max_length=2000)
    proxy: Optional[str] = Field(default=None, max_length=2000)
    api_key: str = Field(..., min_length=1, max_length=4000)


class CredentialUpdate(BaseModel):
    """Update body. All fields optional (partial update). ``api_key`` omitted
    OR empty => keep the existing key (the handler keys off
    ``exclude_unset=True`` + a non-empty check). ``extra='forbid'``."""

    model_config = ConfigDict(extra="forbid")

    name: Optional[str] = Field(default=None, min_length=1, max_length=200)
    description: Optional[str] = Field(default=None, max_length=2000)
    provider: Optional[str] = Field(default=None, min_length=1, max_length=100)
    runtime_scope: Optional[str] = Field(
        default=None, pattern=r"^[a-z][a-z0-9_-]{0,63}$"
    )
    model_name: Optional[str] = Field(
        default=None, min_length=1, max_length=200,
    )
    model_context_tokens: Optional[int] = Field(default=None, gt=0)
    api_url: Optional[str] = Field(default=None, max_length=2000)
    proxy: Optional[str] = Field(default=None, max_length=2000)
    # Empty string is allowed here (means "keep existing"); the handler treats
    # a falsy api_key as "no change".
    api_key: Optional[str] = Field(default=None, max_length=4000)

File: web/src/lib/api/llm-credentials.ts
/**
 * LLM credentials API client for the API Management Center.
 *
 * Thin typed wrappers over the 5 endpoints:
 *
 *   * `POST   /api/v1/llm-credentials`               — create   → OwnerOut
 *   * `GET    /api/v1/llm-credentials`               — list      → PublicOut[]
 *   * `GET    /api/v1/llm-credentials/{id}`          — owner view→ OwnerOut
 *   * `PUT    /api/v1/llm-credentials/{id}`          — update    → OwnerOut
 *   * `DELETE /api/v1/llm-credentials/{id}`          — soft delete (204)
 *
 * Mirrors the hand-rolled-types approach in `mcp-servers.ts` — these routes were
 * added after the last codegen snapshot, so everything funnels through the same
 * per-file `authedFetch` shim (Bearer header + 401 → reopen login).
 *
 * SECURITY: plaintext keys are write-only and never returned by the API. The
 * list surface (`CredentialPublic`) carries id/name/description/provider —
 * NEVER model_name/api_url/api_key. The owner view (`CredentialOwner`) adds
 * model_name/api_url + an `api_key_set` flag, still never the plaintext key.
 */

// ---------------------------------------------------------------------------
// Contract — kept in sync with `routes/llm_credentials.py` +
// `schemas/llm_credentials.py`.
// ---------------------------------------------------------------------------

import type { ResourceAccess } from '@/lib/api/organizations';

/** Public / list view — the surface PromptNode + agent pickers consume. */
export interface CredentialPublic {
  id: string;
  name: string;
  description: string | null;
  provider: string;
  connection_kind?: 'manual' | 'openrouter_oauth';
  runtime_scope: string;
  model_context_tokens: number | null;
  created_at: string;
  updated_at: string;
  access?: ResourceAccess | null;
}

/** Owner management / edit view. NO plaintext key — only `api_key_set`. */
export interface CredentialOwner {
  id: string;
  name: string;
  description: string | null;
  provider: string;
  connection_kind?: 'manual' | 'openrouter_oauth';
  runtime_scope: string;
  model_name: string;
  model_context_tokens: number | null;
  api_url: string | null;
  /** Optional outbound proxy (owner-only — may carry `user:pass@host`). */
  proxy: string | null;
  api_key_set: boolean;
  created_at: string;
  updated_at: string;
  access?: ResourceAccess | null;
}

export interface CreateCredentialBody {
  name: string;
  description?: string | null;
  provider: string;
  /** Runtime catalog that owns this credential. Defaults to Codex. */
  runtime_scope?: string;
  model_name: string;
  model_context_tokens?: number | null;
  api_url?: string | null;
  /** Optional outbound proxy, e.g. `http://host:port`. */
  proxy?: string | null;
  api_key: string;
}

export interface UpdateCredentialBody {
  name?: string;
  description?: string | null;
  provider?: string;
  runtime_scope?: string;
  model_name?: string;
  model_context_tokens?: number | null;
  api_url?: string | null;
  /** Optional outbound proxy, e.g. `http://host:port`. */
  proxy?: string | null;
  /** Omit or send empty to KEEP the existing key. */
  api_key?: string;
}

export interface CredentialConnectionTest {
  ok: boolean;
  outcome: 'connected' | 'credentials_rejected' | 'endpoint_rejected' | 'unreachable';
  latency_ms: number;
  upstream_status: number | null;
}

export interface OpenRouterModel {
  id: string;
  name: string;
  description: string;
  context_length: number | null;
  input_modalities: string[];
  output_modalities: string[];
  supports_tools: boolean;
  supports_web_search: boolean;
  supported_reasoning_efforts: string[];
  default_reasoning_effort: string | null;
  pricing: { prompt?: string | null; completion?: string | null };
  available: boolean;
}

export interface OpenRouterConnection {
  connected: boolean;
  credential_id: string | null;
  models: OpenRouterModel[];
  catalog_refreshed_at: string | null;
  catalog_stale: boolean;
  error_code: string | null;
}

export interface OpenRouterStart {
  authorization_url: string;
  expires_at: string;
}

// ---------------------------------------------------------------------------
// Internal: same `authedFetch` shim as `mcp-servers.ts`.
// ---------------------------------------------------------------------------

import { getApiBase } from '@/lib/base-path';

const BASE = getApiBase();

async function authedFetch(
  path: string,
  init: RequestInit = {},
): Promise<Response> {
  const { useAuthStore } = await import('@/stores/auth');
  const token = useAuthStore.getState().token;
  const headers = new Headers(init.headers);
  if (token) headers.set('Authorization', `Bearer ${token}`);
  if (init.body && !headers.has('Content-Type')) {
    headers.set('Content-Type', 'application/json');
  }
  const resp = await fetch(`${BASE}${path}`, { ...init, headers });
  if (resp.status === 401) {
    useAuthStore.getState().handle401();
  }
  return resp;
}

async function jsonOrThrow<T>(resp: Response, label: string): Promise<T> {
  if (!resp.ok) {
    let detail = '';
    let code: string | null = null;
    try {
      const j = await resp.json();
      if (j && typeof j === 'object' && 'detail' in j) {
        const rawDetail = (j as { detail: unknown }).detail;
        if (typeof rawDetail === 'string') {
          detail = ` — ${rawDetail}`;
        } else if (rawDetail && typeof rawDetail === 'object' && 'code' in rawDetail) {
          code = String((rawDetail as { code: unknown }).code);
          detail = ` — ${code}`;
        }
      }
    } catch {
      // ignore non-JSON bodies
    }
    throw Object.assign(new Error(
      `${label} failed: ${resp.status} ${resp.statusText}${detail}`,
    ), { code });
  }
  return (await resp.json()) as T;
}

// ---------------------------------------------------------------------------
// Public API.
// ---------------------------------------------------------------------------

export async function listLlmCredentials(): Promise<CredentialPublic[]> {
  const resp = await authedFetch('/api/v1/llm-credentials');
  return jsonOrThrow<CredentialPublic[]>(resp, 'listLlmCredentials');
}

export async function getLlmCredential(id: string): Promise<CredentialOwner> {
  const resp = await authedFetch(`/api/v1/llm-credentials/${id}`);
  return jsonOrThrow<CredentialOwner>(resp, 'getLlmCredential');
}

export async function createLlmCredential(
  body: CreateCredentialBody,
): Promise<CredentialOwner> {
  const resp = await authedFetch('/api/v1/llm-credentials', {
    method: 'POST',
    body: JSON.stringify(body),
  });
  return jsonOrThrow<CredentialOwner>(resp, 'createLlmCredential');
}

export async function updateLlmCredential(
  id: string,
  body: UpdateCredentialBody,
): Promise<CredentialOwner> {
  const resp = await authedFetch(`/api/v1/llm-credentials/${id}`, {
    method: 'PUT',
    body: JSON.stringify(body),
  });
  return jsonOrThrow<CredentialOwner>(resp, 'updateLlmCredential');
}

export async function deleteLlmCredential(id: string): Promise<void> {
  const resp = await authedFetch(`/api/v1/llm-credentials/${id}`, {
    method: 'DELETE',
  });
  if (!resp.ok) {
    throw new Error(
      `deleteLlmCredential failed: ${resp.status} ${resp.statusText}`,
    );
  }
}

export async function testLlmCredentialConnection(
  id: string,
): Promise<CredentialConnectionTest> {
  const resp = await authedFetch(`/api/v1/llm-credentials/${id}/test`, {
    method: 'POST',
  });
  return jsonOrThrow<CredentialConnectionTest>(resp, 'testLlmCredentialConnection');
}

export async function getOpenRouterConnection(): Promise<OpenRouterConnection> {
  const resp = await authedFetch('/api/v1/llm-credentials/openrouter/status');
  return jsonOrThrow<OpenRouterConnection>(resp, 'getOpenRouterConnection');
}

export async function startOpenRouterConnection(): Promise<OpenRouterStart> {
  const resp = await authedFetch('/api/v1/llm-credentials/openrouter/start', {
    method: 'POST',
  });
  return jsonOrThrow<OpenRouterStart>(resp, 'startOpenRouterConnection');
}

export async function completeOpenRouterConnection(
  code: string,
  state: string,
): Promise<OpenRouterConnection> {
  const resp = await authedFetch('/api/v1/llm-credentials/openrouter/callback', {
    method: 'POST',
    body: JSON.stringify({ code, state }),
  });
  return jsonOrThrow<OpenRouterConnection>(resp, 'completeOpenRouterConnection');
}

export async function refreshOpenRouterModels(): Promise<OpenRouterConnection> {
  const resp = await authedFetch('/api/v1/llm-credentials/openrouter/refresh', {
    method: 'POST',
  });
  return jsonOrThrow<OpenRouterConnection>(resp, 'refreshOpenRouterModels');
}

export async function disconnectOpenRouter(): Promise<void> {
  const resp = await authedFetch('/api/v1/llm-credentials/openrouter', {
    method: 'DELETE',
  });
  if (!resp.ok) {
    throw new Error(`disconnectOpenRouter failed: ${resp.status} ${resp.statusText}`);
  }
}
